### PR TITLE
Trunk 4755 Implementing Program Enrollment Attributes

### PR DESCRIPTION
**Pull request by jtatia**
migrated Bahmni Code to Core.
link to issue: https://issues.openmrs.org/browse/TRUNK-4755

### CODE REVIEW

**Comment by dkayiwa** [comment]
@jtatia hopefully you are taking a look at the codacy review comments. 😊

**Comment by wluyima** [comment]
Thanks @jtatia ! Can you address the issues pointed out by codacy e.g unused imports, instantation of boolean objects. Normally formatting code and organising imports using the IDE shortcuts should get rid of unused imports

File: api/src/main/java/org/openmrs/PatientProgram.java [hunk 1]
```diff
@@ -10,18 +10,23 @@
...
+import org.openmrs.attribute.Attribute;
+import org.openmrs.customdatatype.CustomValueDescriptor;
+import org.openmrs.customdatatype.Customizable;
 import org.openmrs.util.OpenmrsUtil;
 
 /**
  * PatientProgram
  */
-public class PatientProgram extends BaseOpenmrsData {
+public class PatientProgram extends BaseOpenmrsData implements Customizable<PatientProgramAttribute> {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgram.java]: Should actually extend BaseCustomizableMetadata which will reduce the code changes you made

Comment (wluyima) [reply api/src/main/java/org/openmrs/PatientProgram.java]: Sorry, meant to say BaseCustomizableData, I've updated the comment above to say BaseCustomizableData

File: api/src/main/java/org/openmrs/PatientProgram.java [hunk 2]
```diff
@@ -57,6 +64,11 @@ public PatientProgram() {
 	public PatientProgram(Integer patientProgramId) {
 		setPatientProgramId(patientProgramId);
 	}
+
+	/**Constructor with Patient Program */
+	public PatientProgram(PatientProgram patientProgram) {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgram.java]: Can you please remove this? They had it in Bahmini just because they wanted to construct their own instance from a PatientProgram so they can link them, you don't need it in core.

File: api/src/main/java/org/openmrs/PatientProgram.java [hunk 3]
```diff
@@ -398,4 +410,86 @@ public void setLocation(Location location) {
...
+							this.getAttributes().remove(existing);
+						}
+					}
+				}
+
+				this.getAttributes().add(attribute);
+				attribute.setOwner(this);
+			}
+		}
+	}
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgram.java]: You won't need most of these methods because they are already implemented in BaseCustomizableData, and that's why you need to extend it

File: api/src/main/java/org/openmrs/PatientProgramAttribute.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs;
+
+import org.openmrs.PatientProgram;
+import org.openmrs.attribute.Attribute;
+import org.openmrs.attribute.BaseAttribute;
+
+public class PatientProgramAttribute extends BaseAttribute<ProgramAttributeType, PatientProgram> implements Attribute<ProgramAttributeType, PatientProgram> {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgramAttribute.java]: This needs to be cleaned up to remove the duplication, `BaseAttribute` already implements `Attribute`

Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgramAttribute.java]: MIssing javadocs

Comment (wluyima) [reply api/src/main/java/org/openmrs/PatientProgramAttribute.java]: And you need to add @since annotations

File: api/src/main/java/org/openmrs/PatientProgram.java [hunk 5]
```diff
@@ -398,4 +410,86 @@ public void setLocation(Location location) {
 		Collections.sort(sortedStates);
 		return sortedStates;
 	}
+
+	@Override
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgram.java]: Am starting to think that may be you might want to override these method to add @since annotations but delegate to the super class methods, what do you think @dkayiwa ?

File: api/src/main/java/org/openmrs/PatientProgramAttribute.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    public void setId(Integer id) {
+        setPatientProgramAttributeId(id);
+    }
+
+    public PatientProgram getPatientProgram() {
+        return getOwner();
+    }
+
+    public void setPatientProgram(PatientProgram patientProgram) {
+        setOwner(new PatientProgram(patientProgram));
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/PatientProgramAttribute.java]: This should just set patientProgram without creating another instance

File: api/src/main/java/org/openmrs/ProgramAttributeType.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs;
+
+
+import org.openmrs.attribute.AttributeType;
+import org.openmrs.attribute.BaseAttributeType;
+
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/ProgramAttributeType.java]: Missing javadocs

File: api/src/main/java/org/openmrs/ProgramAttributeType.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+public class ProgramAttributeType extends BaseAttributeType<PatientProgram> implements AttributeType<PatientProgram> {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/ProgramAttributeType.java]: Same here with the duplication, `BaseAttributeType` already implements `AttributeType`

File: api/src/main/java/org/openmrs/ProgramAttributeType.java (hunk 7 above)
Comment (wluyima) [reply api/src/main/java/org/openmrs/ProgramAttributeType.java]: And @since

File: api/src/main/java/org/openmrs/api/ProgramWorkflowService.java [hunk 9]
```diff
@@ -530,4 +533,60 @@ public void purgeConceptStateConversion(ConceptStateConversion conceptStateConve
 	 * @should throw an error when multiple program states with same uuid are found
 	 */
 	public ConceptStateConversion getConceptStateConversionByUuid(String uuid);
+
+	/**
+	 * Fetches a list of all ProgramAttributeTypes
+	 *
+	 * @return list of ProgramAttributeType objects
+	 */
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/ProgramWorkflowService.java]: Missing @since annotations on the new methods

File: api/src/main/java/org/openmrs/api/ProgramWorkflowService.java [hunk 10]
```diff
@@ -530,4 +533,60 @@ public void purgeConceptStateConversion(ConceptStateConversion conceptStateConve
...
+	@Transactional(readOnly = true)
+    @Authorized({"View PatientProgram Attribute Types"})
```
[the remark this answers is not in the document]
Comment (wluyima) [reply api/src/main/java/org/openmrs/api/ProgramWorkflowService.java]: I don't see where you register these new privileges

File: api/src/main/java/org/openmrs/api/ProgramWorkflowService.java [hunk 11]
```diff
@@ -530,4 +533,60 @@ public void purgeConceptStateConversion(ConceptStateConversion conceptStateConve
...
+    List<ProgramAttributeType> getAllProgramAttributeTypes();
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/ProgramWorkflowService.java]: I know they forgot this in Bahmni, but you need to add a `getAllProgramAttributeTypes()` method that takes in boolean argument that specifies whether retired ones should be included or excluded. In theory, this method that has no arguments should delegate to the other with a value of false. I personally prefer naming these methods `getProgramAttributeTypes` rather than `getAllProgramAttributeTypes` but I guess most existing methods are named that way so I guess it's fine.

File: api/src/main/java/org/openmrs/ProgramAttributeType.java (hunk 8 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/ProgramAttributeType.java]: I think this should be `PatientProgramAttributeType`

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateProgramWorkflowDAO.java [hunk 12]
```diff
@@ -295,16 +299,13 @@ public ConceptStateConversion getConceptStateConversionByUuid(String uuid) {
...
 		    "from ConceptStateConversion csc where csc.uuid = :uuid").setString("uuid", uuid).uniqueResult();
 	}
-	
-	/**
-	 * @see org.openmrs.api.db.ProgramWorkflowDAO#getPatientProgramByUuid(java.lang.String)
-	 */
+
 	@Override
 	public PatientProgram getPatientProgramByUuid(String uuid) {
-		return (PatientProgram) sessionFactory.getCurrentSession().createQuery(
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateProgramWorkflowDAO.java]: Would be nice to leave existing code the way it is that is not related to this ticket

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateProgramWorkflowDAO.java [hunk 13]
```diff
@@ -386,4 +387,54 @@ public ProgramWorkflow getWorkflowByUuid(String uuid) {
 		squery.setEntity("concept", concept);
 		return squery.list();
 	}
+
+	/**
+	 * @see org.openmrs.api.db.ProgramWorkflowDAO#getAllProgramAttributeTypes()
+	 */
+	@Override
+	public List<ProgramAttributeType> getAllProgramAttributeTypes() {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateProgramWorkflowDAO.java]: You should add the boolean argument to this method, the one with no arguments shouldn't  exist in the DAO

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 14]
```diff
@@ -572,4 +572,97 @@
 		<modifyDataType tableName="obs" columnName="value_complex" newDataType="varchar(1000)"/>
 	</changeSet>
 
-</databaseChangeLog>
+<changeSet id="TRUNK-4755-20171004-1100" author="jtatia">
+<preConditions onFail="MARK_RAN">
+	<not>
+		<tableExists tableName="program_attribute_type" />
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: I think table name should be `patient_program_attribute_type`

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 15]
```diff
@@ -572,4 +572,97 @@
...
+	<column name="creator" type="int"><constraints nullable="false"/></column>
+	<column name="date_created" type="datetime"><constraints nullable="false"/></column>
+	<column name="changed_by" type="int"/>
+	<column name="date_changed" type="datetime" />
+	<column name="voided" type="boolean" defaultValueBoolean="false">
+		<constraints nullable="false"/>
+	</column>
+	<column name="voided_by" type="int" />
+	<column name="date_voided" type="datetime" />
+	<column name="void_reason" type="varchar(255)" defaultValue="null"/>
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: I think this defaultValue here is redundant

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 16]
```diff
@@ -572,4 +572,97 @@
...
+<addForeignKeyConstraint constraintName="patient_program_attribute_attributetype_fk"
+						 baseTableName="patient_program_attribute" baseColumnNames="attribute_type_id"
+						 referencedTableName="program_attribute_type" referencedColumnNames="program_attribute_type_id" />
+<addForeignKeyConstraint constraintName="patient_program_attribute_creator_fk"
+						 baseTableName="patient_program_attribute" baseColumnNames="creator"
+						 referencedTableName="users" referencedColumnNames="user_id" />
+<addForeignKeyConstraint constraintName="patient_program_attribute_changed_by_fk"
+						 baseTableName="patient_program_attribute" baseColumnNames="changed_by"
+						 referencedTableName="users" referencedColumnNames="user_id" />
+<addForeignKeyConstraint constraintName="patient_program_attribute_retired_by_fk"
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: Shouldn't this be voided rather than retired?

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 17]
```diff
@@ -572,4 +572,97 @@
...
+						 baseTableName="patient_program_attribute" baseColumnNames="retired_by"
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: `baseColumnName` here should also be `voided_by`

File: api/src/main/resources/liquibase-update-to-2.1.xml [hunk 18]
```diff
@@ -572,4 +572,97 @@
...
+		<constraints nullable="false" />
+	</column>
+	<column name="changed_by" type="int" />
+	<column name="date_changed" type="datetime" />
+	<column name="retired" type="smallint" defaultValueNumeric="0">
+		<constraints nullable="false" />
+	</column>
+	<column name="retired_by" type="int(11)" />
+	<column name="date_retired" type="datetime" />
+	<column name="retire_reason" type="varchar(255)" defaultValue="null" />
```
Comment (wluyima) [inline api/src/main/resources/liquibase-update-to-2.1.xml]: Defining a defaultValue of null here is redundant

File: api/src/main/resources/org/openmrs/api/db/hibernate/PatientProgram.hbm.xml [hunk 19]
```diff
@@ -74,6 +74,12 @@
 		<key column="patient_program_id" not-null="true"/>
 		<one-to-many class="org.openmrs.PatientState" />
 	</set>
+
+    <set name="attributes" lazy="true" inverse="true" batch-size="100"
+         cascade="all-delete-orphan" order-by="voided asc">
+        <key column="patient_program_id" />
+        <one-to-many class="org.openmrs.PatientProgramAttribute" />
```
Comment (wluyima) [inline api/src/main/resources/org/openmrs/api/db/hibernate/PatientProgram.hbm.xml]: I think a package is defined at the top of the hbm file, so `PatientProgramAttribute` would be enough.

File: api/src/test/java/org/openmrs/api/impl/ProgramWorkflowServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import static org.mockito.Mockito.verify;
+import org.openmrs.ProgramAttributeType;
+import org.openmrs.api.ProgramWorkflowService;
+import org.openmrs.api.db.ProgramWorkflowDAO;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+@RunWith(PowerMockRunner.class)
+public class ProgramWorkflowServiceImplTest {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/impl/ProgramWorkflowServiceImplTest.java]: These tests need to be in ProgramWorkflowServiceTest, you don't need this new class

File: api/src/test/java/org/openmrs/api/impl/ProgramWorkflowServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+    ProgramWorkflowService programWorkflowService;
+
+    @Mock
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/impl/ProgramWorkflowServiceImplTest.java]: These tests should not be mocked, these are meant to be end to end integration tests to test both the service and dao layers

File: api/src/test/resources/org/openmrs/api/include/ProgramAttributesDataset.xml [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    v. 2.0. If a copy of the MPL was not distributed with this file, You can
+    obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+    the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+
+    Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+    graphic logo is a trademark of OpenMRS Inc.
+
+-->
+<dataset>
+    <program_attribute_type program_attribute_type_id="1" name="ProgramId" min_occurs="1" retired="false"
```
Comment (wluyima) [inline api/src/test/resources/org/openmrs/api/include/ProgramAttributesDataset.xml]: I don't see where you are using this file

Comment (wluyima) [reply api/src/test/resources/org/openmrs/api/include/ProgramAttributesDataset.xml]: In theory, you should use it for the integration tests in `ProgramWorkflowServiceTest` after you get rid of the mocks

File: api/src/main/java/org/openmrs/PatientProgram.java (hunk 5 above)
Comment (jtatia) [reply api/src/main/java/org/openmrs/PatientProgram.java]: any decision on this. @dkayiwa ?

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/PatientProgram.java]: I agree with @wluyima

**Comment by dkayiwa** [comment]
@jtatia did you look at the codacy review comments?

**Comment by jtatia** [comment]
Yes @dkayiwa. This PR is not ready for review. Still need to make a few more changes. I'll try to get it done by tomorrow.

**Comment by dkayiwa** [comment]
@jtatia did you see the merge conflicts on this pull request? 😊

**Comment by dkayiwa** [comment]
@jtatia are you still working on this?
